hal/spi_api.h

```diff
@@ -19,220 +18,346 @@
+ * @return SPI_RESULT_OK on success. See spi_result_t for more details about failures.
  */
-int spi_master_block_write(spi_t *obj, const char *tx_buffer, int tx_length, char *rx_buffer, int rx_length, char write_fill);
+spi_result_t spi_init(spi_t *obj, const spi_init_t *init);
 
-/** Check if a value is available to read
+/**
+ * Processes a transfer blocking until completion if the device is available.
  *
- * @param[in] obj The SPI peripheral to check
- * @return non-zero if a value is available
- */
-int  spi_slave_receive(spi_t *obj);
-
-/** Get a received value out of the SPI receive buffer in slave mode
+ * - returns `true` if the operation succeeded.
+ * - returns `false` if any of the parameters is NULL.
+ * - returns `false` if the peripheral is busy with another transaction.
+ * - returns `false` if an error occurred. An error can be due to :
+ *   - conflicting master ;
+ *   - rx or tx overflow ;
+ *   - rx or tx underflow...
+ * - sends and received up to `max(tx_count, rx_count)`.
+ * - if `tx` is not NULL then writes up to `tx_count` symbol from `tx` then sends
+ *   `max(tx_count, rx_count) - tx_count` times the `fill_symbol` .
+ * - if `rx` is not NULL then stores in `rx` up to `rx_count` symbol and continue popping symbol
+ *   from the peripheral until `max(tx_count, rx_count) - rx_count` is reached. All spare symbol
+ *   are dropped.
+ * - if the device is available, it locks the device and blocks until completion (or error).
  *
- * Blocks until a value is available
- * @param[in] obj The SPI peripheral to read
- * @return The value received
- */
-int  spi_slave_read(spi_t *obj);
-
-/** Write a value to the SPI peripheral in slave mode
+ * @warning The spi_transfer_args_t instance must live for the whole duration of this method.
  *
- * Blocks until the SPI peripheral can be written to
- * @param[in] obj   The SPI peripheral to write
- * @param[in] value The value to write
- */
-void spi_slave_write(spi_t *obj, int value);
-
-/** Checks if the specified SPI peripheral is in use
+ * @param[in,out] obj   An initialised spi_t instance.
+ * @param[in]     args  An spi_transfer_args_t instance.
  *
- * @param[in] obj The SPI peripheral to check
- * @return non-zero if the peripheral is currently transmitting
+ * @return True on success.
  */
-int  spi_busy(spi_t *obj);
+bool spi_transfer(spi_t *obj, const spi_transfer_args_t *args);
 
-/** Get the module number
+/**
+ * Frees the SPI instance.
+ *
+ * - does nothing if passed a NULL pointer.
+ * - cancel all pending or on going transaction/transfer and then deinitialise (clock out) this
+ *   device if it was the last device associated to this peripheral.
  *
- * @param[in] obj The SPI peripheral to check
- * @return The module number
+ * @param[in,out] obj   An initialised spi_t instance.
+ *
+ * @return `true` on success.
  */
-uint8_t spi_get_module(spi_t *obj);
-
-/**@}*/
+void spi_free(spi_t *obj);
 
-#if DEVICE_SPI_ASYNCH
+#if DEVICE_SPISLAVE
```

Comment:
SPI slave mode will need some event or callback to indicate that the master is transferring data if you want the CPU to be able to sleep. Otherwise a thread will need to be always busy waiting for the next packet.

Reply:
This will be part of the ASYNC API (in spi_async.h yet to be completed).
The functions here are provided to fit the current SPISlave class.

Reply:
I would recommend removing this rather than deprecating it. This is allowed at the HAL layer as long as it isn't a breaking change to user. This should be a breaking change as long as SPISlave came be implemented with another mechanism (async).